packages/aws-cdk/lib/cdk-toolkit.ts

```diff
@@ -243,6 +248,82 @@ export class CdkToolkit {
     }
   }
 
+  public async watch(options: WatchOptions) {
+    const rootDir = path.dirname(path.resolve(PROJECT_CONFIG));
+    debug("root directory used for 'watch' is: %s", rootDir);
+
+    const watchSettings: { include?: string | string[], exclude: string | string [] } | undefined =
+        this.props.configuration.settings.get(['watch']);
+    if (!watchSettings) {
+      throw new Error("Cannot use the 'watch' command without specifying at least one directory to monitor. " +
+        'Make sure to add a "watch" key to your cdk.json');
+    }
+
+    // For the "include" subkey under the "watch" key, the behavior is:
+    // 1. No "watch" setting? We error out.
+    // 2. "watch" setting without an "include" key? We default to observing "./**".
+    // 3. "watch" setting with an empty "include" key? We default to observing "./**".
+    // 4. Non-empty "include" key? Just use the "include" key.
+    const watchIncludes = this.patternsArrayForWatch(watchSettings.include, rootDir, /* returnRootDirIfEmpty */ true);
+    debug("'include' patterns for 'watch': %s", watchIncludes);
+
+    // For the "exclude" subkey under the "watch" key,
+    // the behavior is to add some default excludes in addition to the ones specified by the user:
+    // 1. The CDK output directory.
+    // 2. Any file whose name starts with a dot.
+    // 3. Any directory's content whose name starts with a dot.
+    // 4. Any node_modules and its content.
+    const outputDir = this.props.configuration.settings.get(['output']);
+    const watchExcludes = this.patternsArrayForWatch(watchSettings.exclude, rootDir, /* returnRootDirIfEmpty */ false).concat(
+      `${rootDir}/${outputDir}/**`,
+      `${rootDir}/**/.*`,
+      `${rootDir}/**/.*/**`,
+      `${rootDir}/**/node_modules/**`,
+    );
+    debug("'exclude' patterns for 'watch': %s", watchExcludes);
+
+    // Since 'cdk deploy' is a relatively slow operation for a 'watch' process,
+    // introduce a concurrency latch that tracks the state.
+    // This way, if file change events arrive when a 'cdk deploy' is still executing,
+    // we will batch them, and trigger another 'cdk deploy' after the current one finishes,
+    // making sure 'cdk deploy's  always execute one at a time.
+    let latch: 'pre-ready' | 'open' | 'deploying' | 'queued' = 'pre-ready';
+    chokidar.watch(watchIncludes, {
+      ignored: watchExcludes,
+      cwd: rootDir,
+      // ignoreInitial: true,
+    }).on('ready', () => {
+      latch = 'open';
+      debug("'watch' received the 'ready' event. From now on, all file changes will trigger a deployment");
+    }).on('all', async (event, filePath) => {
+      if (latch === 'pre-ready') {
```

Comment:
Do we really need `pre-ready`? Can't we just add the `all` handler inside the `ready` handler?

```ts
let latch = 'open';
const watch = chokidar.watch(...);
watch.on('ready', () => {
  watch.on('all', async (e, f) => {
    switch (latch) {
      /...
  });
});
```

Reply:
If we install the `'all'` event handler only after the `'ready'` event has fired, then we won't be able to print out the starting messages (where we say "watch is observing the file/directory xyz" for changes").

Reply:
Why not? Can't we print this message after `ready` is fired?

Reply:
No, because if we don't install the handler before the `'ready'` event is fired, we won't be notified for the initial events that start observing the given files for changes.

I'm not saying we _have to_ have those output, but I feel like it's nice to show the customer which files we are observing for changes 🙂.

Reply:
OK, can't say I fully understand this, but it's not that important. Just thought it might be nicer to simplify.